specifyweb/frontend/js_src/lib/components/AttachmentsBulkImport/AttachmentsUpload.tsx

```diff
@@ -0,0 +1,428 @@
+import type { SerializedResource } from '../DataModel/helperTypes';
+import type { SpecifyResource } from '../DataModel/legacyTypes';
+import { schema, strictGetModel } from '../DataModel/schema';
+import type { Attachment, Tables } from '../DataModel/types';
+import { Dialog, LoadingScreen } from '../Molecules/Dialog';
+import type { AttachmentUploadSpec, EagerDataSet } from './Import';
+import { PerformAttachmentTask } from './PerformAttachmentTask';
+import type {
+  AttachmentStatus,
+  AttachmentWorkStateProps,
+  PartialUploadableFileSpec,
+  UploadAttachmentSpec,
+} from './types';
+import {
+  getAttachmentsFromResource,
+  resolveAttachmentRecord,
+  validateAttachmentFiles,
+} from './utils';
+
+async function prepareForUpload(
+  dataSet: EagerDataSet,
+  baseTableName: keyof Tables
+): Promise<RA<PartialUploadableFileSpec>> {
+  const validatedFiles = await validateAttachmentFiles(
+    dataSet.uploadableFiles,
+    dataSet.uploadSpec as AttachmentUploadSpec,
+    // If user validated before, and chose disambiguation, need to preserve it
+    true
+  );
+  const mappedUpload = await Promise.all(
+    validatedFiles.map(async (uploadable) =>
+      uploadFileWrapped(uploadable, baseTableName, undefined, true)
+    )
+  );
+  const fileNamesToTokenize = filterArray(
+    mappedUpload.map((uploadable) =>
+      uploadable.status?.type === 'matched' &&
+      uploadable.uploadTokenSpec === undefined
+        ? uploadable.file.name
+        : undefined
+    )
+  );
+  if (fileNamesToTokenize.length === 0) return mappedUpload;
+  return ajax<RA<UploadAttachmentSpec>>('/attachment_gw/get_upload_params/', {
+    method: 'POST',
+    headers: { Accept: 'application/json' },
+    body: {
+      filenames: fileNamesToTokenize,
+    },
+  }).then(({ data }) => {
+    if (fileNamesToTokenize.length !== data.length) {
+      // Throwing an error for development testing. Hasn't happened yet.
+      throw new Error(
+        'DEV: length changed in between effect calls. Unsafe. Aborting upload.'
+      );
+    }
+    let indexInTokenData = 0;
+    return mappedUpload.map((uploadableFile) => {
+      let uploadToken: UploadAttachmentSpec | undefined = undefined;
+      if (
+        uploadableFile.status?.type === 'matched' &&
+        uploadableFile.uploadTokenSpec === undefined
+      ) {
+        uploadToken = data[indexInTokenData];
+        indexInTokenData += 1;
+      }
+      return {
+        ...uploadableFile,
+        uploadTokenSpec: uploadToken ?? uploadableFile.uploadTokenSpec,
+      };
+    });
+  });
+}
+
+export function SafeUploadAttachmentsNew({
+  dataSet,
+  baseTableName,
+  onSync: handleSync,
+}: {
+  readonly dataSet: EagerDataSet;
+  readonly onSync: (
+    generatedState: RA<PartialUploadableFileSpec> | undefined,
+    isSyncing: boolean
+  ) => void;
+  readonly baseTableName: keyof Tables;
+}): JSX.Element {
+  const uploadDisabled = React.useMemo(() => dataSet.needsSaved, [dataSet]);
```

Comment:
nitpick: dependency array could be `[dataSet.needsSaved]` instead of `[dataSet]`

Reply:
Simplified